source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -351,29 +453,78 @@ void RedisCluster::RedisDiscoverySession::onResponse(
       return;
     }
 
-    // Field 2: Primary address for slot range
-    auto primary_address = ProcessCluster(slot_range[SlotPrimary]);
-    if (!primary_address) {
+    // Row 3: Primary slot address
+    if (!validateCluster(slot_range[SlotPrimary])) {
       onUnexpectedResponse(value);
       return;
     }
+    // Try to parse primary slot address as IP address
+    // It may fail in AWS ElastiCache use case: it uses hostnames instead of IPs.
+    // If this is the case - we'll come back later and try to resolve hostnames asynchronously.
+    ClusterSlot slot(slot_range[SlotRangeStart].asInteger(), slot_range[SlotRangeEnd].asInteger(),
+                     processClusterByIP(slot_range[SlotPrimary]));
+    if (slot.primary() == nullptr) {
+      // Primary address is hostname: save the name for further resolving
+      const auto& array = slot_range[SlotPrimary].asArray();
+      slot.primary_hostname_ = array[0].asString();
+      slot.primary_port_ = array[1].asInteger();
+    }
 
-    slots->emplace_back(slot_range[SlotRangeStart].asInteger(),
-                        slot_range[SlotRangeEnd].asInteger(), primary_address);
-
+    // Row 4-N: Replica(s) addresses
     for (auto replica = std::next(slot_range.begin(), SlotReplicaStart);
          replica != slot_range.end(); ++replica) {
-      auto replica_address = ProcessCluster(*replica);
-      if (!replica_address) {
+      if (!validateCluster(*replica)) {
         onUnexpectedResponse(value);
         return;
       }
-      slots->back().addReplica(std::move(replica_address));
+      auto replica_address = processClusterByIP(*replica);
+      if (replica_address) {
+        slot.addReplica(std::move(replica_address));
+      } else {
+        // Possible AWS ElastiCache use case: hostname instead of IP
+        const auto& array = replica->asArray();
+        slot.addReplicaToResolve(array[0].asString(), array[1].asInteger());
+      }
     }
+    // If at least one (primary, replicas) address is hostname, schedule DNS resolving
+    if (slot.primary() == nullptr || !slot.replicas_to_resolve_.empty()) {
+      address_resolve_required = true;
+    }
+    cluster_slots->push_back(std::move(slot));
   }
 
-  parent_.onClusterSlotUpdate(std::move(slots));
-  resolve_timer_->enableTimer(parent_.cluster_refresh_rate_);
+  if (!address_resolve_required) {
+    // All slots addresses were represented by IP/Port pairs.
+    parent_.onClusterSlotUpdate(std::move(cluster_slots));
+    resolve_timer_->enableTimer(parent_.cluster_refresh_rate_);
+  } else {
+    // Resolve hostnames, once completed run onClusterSlotUpdate() / enable timer
```

Comment:
I'd maybe write this as `// DNS resolution is required, defer finalizing the slot update until resolution is complete.`